torchtune/training/_compile.py

```diff
@@ -5,38 +5,49 @@
 # LICENSE file in the root directory of this source tree.
 
 import os
+from typing import Union
 
 import torch
 from torch import nn
 
-from torchtune.modules import TransformerDecoder, TransformerSelfAttentionLayer
+from torchtune.modules import (
+    DeepFusionModel,
+    TransformerCrossAttentionLayer,
+    TransformerDecoder,
+    TransformerSelfAttentionLayer,
+)
 from torchtune.modules.loss import CEWithChunkedOutputLoss
 from torchtune.utils import get_logger, torch_version_ge
 
 log = get_logger("INFO")
 
 
 def compile_model(
-    model: TransformerDecoder,
+    model: Union[TransformerDecoder, DeepFusionModel],
     verbose: bool = True,
 ) -> None:
     """
     Utility to compile a transformer model inplace. On PyTorch nightlies we use per-layer compile
     to reduce compile times. Otherwise we compile the full model, which takes longer.
 
     Args:
-        model (TransformerDecoder): A transformer model to compile.
+        model (Union[TransformerDecoder, DeepFusionModel]): A model to compile.
+            Can be a TransformerDecoder or DeepFusionModel; in the latter case only
+            the model's decoder will be compiled.
         verbose (bool): Whether to log compile info. Default: True
     Returns:
         None
 
     """
     backend = os.environ.get("TORCH_COMPILE_BACKEND", "inductor")
+    model_to_compile = model if isinstance(model, TransformerDecoder) else model.decoder
```

Comment:
nit: I'd rather just keep it as model with a patch "if isinstance(DeepFusionModel): model = model.decoder"

Reply:
The inverse logic of checking if Deep Fusion model seems tighter